server/src/internalClusterTest/java/org/elasticsearch/get/GetFromTranslogActionIT.java

```diff
@@ -0,0 +1,102 @@
+        assertAcked(
+            prepareCreate("test").setMapping("field1", "type=keyword,store=true")
+                .setSettings(
+                    Settings.builder()
+                        .put("index.refresh_interval", -1)
+                        // A GetFromTranslogAction runs only Stateless where there is only one active indexing shard.
+                        .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                        .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                )
+                .addAlias(new Alias("alias").writeIndex(randomFrom(true, false, null)))
+        );
+        ensureGreen();
+
+        var response = getFromTranslog(indexOrAlias(), "1");
+        assertNull(response.getResult());
+        // There hasn't been any switches from unsafe to safe map
+        assertThat(response.segmentGeneration(), equalTo(-1L));
+
+        client().prepareIndex("test").setId("1").setSource("field1", "value1").setRefreshPolicy(RefreshPolicy.NONE).get();
+        response = getFromTranslog(indexOrAlias(), "1");
+        assertNotNull(response.getResult());
+        assertThat(response.getResult().isExists(), equalTo(true));
+        assertThat(response.segmentGeneration(), equalTo(-1L));
+        // Get followed by a delete should still return a result
+        client().prepareDelete("test", "1").get();
+        response = getFromTranslog(indexOrAlias(), "1");
+        assertNotNull(response.getResult());
+        assertThat(response.getResult().isExists(), equalTo(false));
+        assertThat(response.segmentGeneration(), equalTo(-1L));
+
+        var indexResponse = client().prepareIndex("test").setSource("field1", "value2").get();
+        response = getFromTranslog(indexOrAlias(), indexResponse.getId());
+        assertNotNull(response.getResult());
+        assertThat(response.getResult().isExists(), equalTo(true));
+        assertThat(response.segmentGeneration(), equalTo(-1L));
+        // After a refresh we should not be able to get from translog
+        client().admin().indices().refresh(new RefreshRequest("test")).get();
+        response = getFromTranslog(indexOrAlias(), indexResponse.getId());
+        assertNull(response.getResult());
+        assertThat(response.segmentGeneration(), equalTo(-1L));
+        // After two refreshes the LiveVersionMap switches back to append-only and stops tracking IDs
+        // Refreshing with empty LiveVersionMap doesn't cause the switch, see {@link LiveVersionMap.Maps#shouldInheritSafeAccess()}.
+        client().prepareIndex("test").setSource("field1", "value3").get();
+        client().admin().indices().refresh(new RefreshRequest("test")).get();
+        client().admin().indices().refresh(new RefreshRequest("test")).get();
+        // An optimized index operation marks the maps as unsafe
+        client().prepareIndex("test").setSource("field1", "value4").get();
+        response = getFromTranslog(indexOrAlias(), "non-existent");
```

Comment:
I am not sure I get the whole logic (because I've missed reviewing the other PRs), but I wonder how a non-existent value returns a generation to wait for?